contracts/peripheral/YieldStreamer.sol

```diff
@@ -0,0 +1,466 @@
+        @param recipient_ Address to direct staking yield and vault shares to.
+        @param paymentInterval_ How much time must elapse before yield is able to be swapped for DAI.
+    */
+    function deposit(
+        uint256 amount_,
+        address recipient_,
+        uint256 paymentInterval_,
+        uint256 userMinimumDaiThreshold_
+    ) external override {
+        require(!depositDisabled, "Deposits currently disabled");
+        require(amount_ > 0, "Invalid deposit amount");
+        require(recipient_ != address(0), "Invalid recipient address");
+        require(userMinimumDaiThreshold_ >= minimumDaiThreshold, "minimumDaiThreshold too low");
+
+        IERC20(sOHM).safeTransferFrom(msg.sender, address(this), amount_);
+
+        userDepositIds[msg.sender].push(idCount.current());
+        activeDepositIds.push(idCount.current());
+
+        depositInfo[idCount.current()] = DepositInfo({
+            id: idCount.current(),
+            depositor: msg.sender,
+            recipient: recipient_,
+            principalAmount: amount_,
+            agnosticAmount: _toAgnostic(amount_),
+            lastUpkeepTimestamp: block.timestamp,
+            paymentInterval: paymentInterval_,
+            unclaimedDai: 0,
+            userMinimumDaiThreshold: userMinimumDaiThreshold_
+        });
+
+        idCount.increment();
+
+        emit Deposited(msg.sender, amount_);
+    }
+
+    /**
+        @notice Add more sOHM to your principal deposit.
+        @param id_ Id of the deposit.
+        @param amount_ Amount of sOHM to withdraw.
+    */
+    function addToDeposit(uint256 id_, uint256 amount_) external override {
+        require(!depositDisabled, "Deposits currently disabled");
+        require(amount_ > 0, "Invalid deposit amount");
+
+        DepositInfo storage userDeposit = depositInfo[id_];
+        require(userDeposit.depositor == msg.sender, "Deposit is not yours");
+
+        IERC20(sOHM).safeTransferFrom(msg.sender, address(this), amount_);
+        userDeposit.principalAmount += amount_;
+        userDeposit.agnosticAmount += _toAgnostic(amount_);
+
+        emit Deposited(msg.sender, amount_);
+    }
+
+    /**
+        @notice Withdraw part of the principal amount deposited.
+        @dev Does not allow all the principal to be withdrawn. If you would like to do that use withdrawAll(). Reason is because we want to delete the element in the active deposits array after withdrawing all the principal.
+        @param id_ Id of the deposit.
+        @param amount_ Amount of sOHM to withdraw.
+    */
+    function withdrawPrincipal(uint256 id_, uint256 amount_) external override {
+        require(!withdrawDisabled, "Withdraws currently disabled");
+        require(amount_ > 0, "Invalid withdraw amount");
+
+        DepositInfo storage userDeposit = depositInfo[id_];
+        require(userDeposit.depositor == msg.sender, "Deposit is not yours");
+        require(amount_ < userDeposit.principalAmount, "input >= principal");
```

Comment:
Im doing this not just for that but so amount cant equal principalAmount. If user wants to withdraw all their principal I want them to use withdrawAll() so that internally I can do the cleanup of their deposit from the activeDepositIds. So i was intending this method only for partial withdrawals of principal.

I could allow them to withdraw it all here and leave the yield and unclaimed dai in but that doesn't seem like they would want to do that intentionally.

Im thinking now to do a if statement if amount = principal then just call withdrawAll() instead. They might not expect withdrawing their yield as well and unclaimed dai but that seems better for the user since they would want to do that anyway.